### PR TITLE
BUG: algos.diff with datetimelike and NaT

### PR DESCRIPTION
**Pull request by jbrockmendel**
- [ ] closes #xxxx
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

### CODE REVIEW

File: pandas/_libs/algos.pyx [hunk 1]
```diff
@@ -1231,15 +1237,31 @@ def diff_2d(
...
+                                out[i, j] = left - right
                 else:
                     if periods >= 0:
                         start, stop = periods, sy
                     else:
                         start, stop = 0, sy + periods
                     for j in range(start, stop):
                         for i in range(sx):
-                            out[i, j] = arr[i, j] - arr[i, j - periods]
+                            left = arr[i, j]
```
Comment (jreback) [inline pandas/_libs/algos.pyx:1256]: rather than repeating can you make an in line function?

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
where is the bug?

File: pandas/_libs/algos.pyx (hunk 1 above)
Comment (jbrockmendel) [reply pandas/_libs/algos.pyx:1256]: i wasnt able to, no.  the trouble is we dont have a good way to declare the return type

**Comment by jbrockmendel** [comment]
> where is the bug?

`algos.diff`.  the tests this adds fails in master

File: pandas/core/algorithms.py [hunk 2]
```diff
@@ -1908,6 +1908,8 @@ def diff(arr, n: int, axis: int = 0, stacklevel=3):
 
     if is_extension_array_dtype(dtype):
         if hasattr(arr, f"__{op.__name__}__"):
+            if axis != 0:
+                raise ValueError(f"cannot diff {type(arr).__name__} on axis={axis}")
```
Comment (jreback) [inline pandas/core/algorithms.py:1912]: does this have a test that hits?

File: pandas/tests/test_algos.py [hunk 3]
```diff
@@ -2405,3 +2405,20 @@ def test_index(self):
             dtype="timedelta64[ns]",
         )
         tm.assert_series_equal(algos.mode(idx), exp)
+
+
+class TestDiff:
+    def test_diff_datetimelike_nat(self):
+        # NaT - NaT is NaT, not 0
```
Comment (jreback) [inline pandas/tests/test_algos.py:2412]: can you also test with m8

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/algorithms.py (hunk 2 above)
Comment (jbrockmendel) [reply pandas/core/algorithms.py:1912]: no, will add one

File: pandas/tests/test_algos.py (hunk 3 above)
Comment (jbrockmendel) [reply pandas/tests/test_algos.py:2412]: sure, just pushed

File: pandas/_libs/algos.pyx [hunk 4]
```diff
@@ -1231,15 +1237,31 @@ def diff_2d(
                         start, stop = 0, sx + periods
                     for j in range(sy):
                         for i in range(start, stop):
-                            out[i, j] = arr[i, j] - arr[i - periods, j]
+                            left = arr[i, j]
+                            right = arr[i - periods, j]
+                            if out_t is int64_t and datetimelike:
```
Comment (jreback) [inline pandas/_libs/algos.pyx:1242]: you could make this simpler by
```
if out_t is int64_t and datetimelike and (left == NPY_NAT or right == NPY_NAT):
   .....
else
   ...
```
to make slightly shorter, not a big deal though

File: pandas/_libs/algos.pyx [hunk 5]
```diff
@@ -1218,6 +1221,9 @@ def diff_2d(
     elif (out_t is float64_t
           and (diff_t is float32_t or diff_t is int8_t or diff_t is int16_t)):
         raise NotImplementedError
+    elif out_t is int64_t and diff_t is not int64_t:
```
Comment (jreback) [inline pandas/_libs/algos.pyx:1224]: is this hit by a test (or indirectly by a dispatch)?

Comment (jbrockmendel) [reply pandas/_libs/algos.pyx:1224]: no, this should never be reached (like the line above)

Comment (jreback) [reply pandas/_libs/algos.pyx:1224]: ahh so these are guards, okie

File: pandas/_libs/algos.pyx (hunk 4 above)
Comment (jbrockmendel) [reply pandas/_libs/algos.pyx:1242]: i find this clearer

**Review by jreback — APPROVED** [review]

**Pull request merged**